src/processing/spot.rs

```diff
@@ -217,53 +218,83 @@ pub async fn process_long_tail_asset(
+    pair: &str,
+    source: &str,
+    decimals: u32,
+) -> Result<(u64, f64), MonitoringError> {
     let mut conn = pool.get().await.map_err(MonitoringError::Connection)?;
 
     let config = get_config(None).await;
-    let network_env = &config.network_str();
-    let data_type = "spot";
 
-    let decimals = *config.decimals(DataType::Spot).get(&pair).unwrap();
+    let filtered_by_source_result: Result<SpotEntry, _> = match config.network().name {
+        NetworkName::Testnet => {
+            testnet_dsl::spot_entry
+                .filter(testnet_dsl::pair_id.eq(pair))
+                .filter(testnet_dsl::source.eq(source))
+                .order(testnet_dsl::block_timestamp.desc())
+                .first(&mut conn)
+                .await
+        }
+        NetworkName::Mainnet => {
+            mainnet_dsl::mainnet_spot_entry
+                .filter(mainnet_dsl::pair_id.eq(pair))
+                .filter(mainnet_dsl::source.eq(source))
+                .order(mainnet_dsl::block_timestamp.desc())
+                .first(&mut conn)
+                .await
+        }
+    };
 
-    let mut prices = Vec::new();
-    for src in &sources {
-        let result: Result<SpotEntry, _> = match config.network().name {
-            NetworkName::Testnet => {
-                testnet_dsl::spot_entry
-                    .filter(testnet_dsl::pair_id.eq(&pair))
-                    .filter(testnet_dsl::source.eq(src))
-                    .order(testnet_dsl::block_timestamp.desc())
-                    .first(&mut conn)
-                    .await
-            }
-            NetworkName::Mainnet => {
-                mainnet_dsl::mainnet_spot_entry
-                    .filter(mainnet_dsl::pair_id.eq(&pair))
-                    .filter(mainnet_dsl::source.eq(src))
-                    .order(mainnet_dsl::block_timestamp.desc())
-                    .first(&mut conn)
-                    .await
-            }
-        };
-
-        if let Ok(data) = result {
+    match filtered_by_source_result {
+        Ok(data) => {
+            let time = time_since_last_update(&data);
             let price_as_f64 = data.price.to_f64().ok_or(MonitoringError::Price(
                 "Failed to convert price to f64".to_string(),
             ))?;
             let normalized_price = price_as_f64 / (10_u64.pow(decimals)) as f64;
-            prices.push((src.clone(), normalized_price));
-        }
-    }
-
-    // Calculate deviations between sources
-    for (i, (src1, price1)) in prices.iter().enumerate() {
-        for (src2, price2) in prices.iter().skip(i + 1) {
-            let deviation = (price1 - price2).abs() / price1;
-            LONG_TAIL_ASSET_DEVIATION
-                .with_label_values(&[network_env, &pair, data_type, src1, src2])
-                .set(deviation);
+            let (source_deviation, _) = source_deviation(&data, normalized_price).await?;
+            Ok((time, source_deviation))
         }
+        Err(e) => Err(e.into()),
```

Comment:
can't we just use question mark or `.map_err` in the assignment thus avoiding this `match` operator

Reply:
ah yeah just copied the already existing code - updated that everywhere